File: Src/PathContext.cpp
/**
 * @file  PathContext.cpp
 *
 * @brief Implementation file for PathInfo and PathContext
 *
 */

#include "pch.h"
#include "PathContext.h"
#include <cassert>
#include "paths.h"

/**
 * @brief Copy constructor.
 */
PathInfo::PathInfo(const PathInfo &pi)
: m_sPath(pi.m_sPath)
{
}

/**
 * @brief Get path.
 * @param [in] sbNormalized true if path is wanted in normalized format.
 */
String PathInfo::GetPath(bool bNormalized /*= true*/) const
{ 
	if (!bNormalized)
		return paths::AddTrailingSlash(m_sPath);
	else
		return m_sPath;
}

/**
 * @brief Normalize path.
 */
void PathInfo::NormalizePath()
{
	paths::normalize(m_sPath);
}

PathContext::PathContext()
{
	m_nFiles = 0;
}

PathContext::PathContext(const String& sLeft)
{
	m_nFiles = 1;
	m_path[0].SetPath(sLeft);
}

PathContext::PathContext(const String& sLeft, const String& sRight)
{
	m_nFiles = 2;
	m_path[0].SetPath(sLeft);
	m_path[1].SetPath(sRight);
}

PathContext::PathContext(const String& sLeft, const String& sMiddle, const String& sRight)
{
	m_nFiles = 3;
	m_path[0].SetPath(sLeft);
	m_path[1].SetPath(sMiddle);
	m_path[2].SetPath(sRight);
}

PathContext::PathContext(const PathContext &paths)
{
	m_nFiles = paths.m_nFiles;
	std::copy_n(paths.m_path, m_nFiles, m_path);
}

PathContext::PathContext(const std::vector<String> &paths)
{
	m_nFiles = static_cast<int>(paths.size());
	for (size_t i = 0; i < paths.size(); i++)
		m_path[i].SetPath(paths[i]);
}

String PathContext::GetAt(int nIndex) const
{
	assert(nIndex < m_nFiles);
	return m_path[nIndex].GetPath();
}

String& PathContext::GetElement(int nIndex)
{
	assert(nIndex < m_nFiles);
	return m_path[nIndex].GetRef();
}

void PathContext::SetAt(int nIndex, const String& newElement)
{
	assert(nIndex < m_nFiles);
	m_path[nIndex].SetPath(newElement);
}

/**
 * @brief Empty m_path array
 */
void PathContext::RemoveAll()
{
	m_nFiles = 0;
	m_path[0].SetPath(_T(""));
	m_path[1].SetPath(_T(""));
	m_path[2].SetPath(_T(""));
}

/**
 * @brief Return left path.
 * @param [in] sNormalized If true normalized path is returned.
 */
String PathContext::GetLeft(bool bNormalized) const
{
	if (m_nFiles == 0)
		return _T("");
	return m_path[0].GetPath(bNormalized);
}

/**
 * @brief Return right path.
 * @param [in] sNormalized If true normalized path is returned.
 */
String PathContext::GetRight(bool bNormalized) const
{
	if (m_nFiles < 2)
		return _T("");
	return m_path[m_nFiles - 1].GetPath(bNormalized);
}

/**
 * @brief Return middle path.
 * @param [in] sNormalized If true normalized path is returned.
 */
String PathContext::GetMiddle(bool bNormalized) const
{
	if (m_nFiles < 3)
		return _T("");
	return m_path[1].GetPath(bNormalized);
}

/**
 * @brief Return path
 * @param [in] index index of path to return
 * @param [in] sNormalized If true normalized path is returned.
 */
String PathContext::GetPath(int index, bool bNormalized) const
{
	return m_path[index].GetPath(bNormalized);
}

/**
 * @brief Set left path.
 * @param [in] path New path for item.
 */
void PathContext::SetLeft(const String& path, bool bNormalized)
{
	if (m_nFiles == 0)
		m_nFiles = 1;
	m_path[0].SetPath(path);
	if (bNormalized)
		m_path[0].NormalizePath();
}

/**
 * @brief Set right path.
 * @param [in] path New path for item.
 */
void PathContext::SetRight(const String& path, bool bNormalized)
{
	if (m_nFiles < 2)
		m_nFiles = 2;
	m_path[m_nFiles - 1].SetPath(path);
	if (bNormalized)
		m_path[m_nFiles - 1].NormalizePath();
}

/**
 * @brief Set middle path.
 * @param [in] path New path for item.
 */
void PathContext::SetMiddle(const String& path, bool bNormalized)
{
	if (m_nFiles < 3)
	{
		m_nFiles = 3;
		m_path[2] = m_path[1];
	}
	m_path[1].SetPath(path);
	if (bNormalized)
		m_path[1].NormalizePath();
}

/**
 * @brief Set path
 * @param [in] index index of path to set
 * @param [in] path New path for item.
 */
void PathContext::SetPath(int index, const String& path, bool bNormalized)
{
	if (index >= sizeof(m_path)/sizeof(m_path[0]))
		return;
	if (index >= m_nFiles)
		m_nFiles = index + 1;
	m_path[index].SetPath(path);
	if (bNormalized)
		m_path[index].NormalizePath();
}

/**
 * @brief Swap paths.
 */
void PathContext::Swap()
{
	if (m_nFiles < 3)
		m_path[0].m_sPath.swap(m_path[1].m_sPath);
	else
		m_path[0].m_sPath.swap(m_path[2].m_sPath);
}

File: winmerge-core/src/main/java/org/winmerge/core/io/PathContext.java
package org.winmerge.core.io;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public final class PathContext implements Iterable<String> {
    private int files;
    private final PathInfo[] path;

    public PathContext() {
        files = 0;
        path = new PathInfo[] {new PathInfo(), new PathInfo(), new PathInfo()};
    }

    public PathContext(String left) {
        this();
        files = 1;
        path[0].setPath(left);
    }

    public PathContext(String left, String right) {
        this();
        files = 2;
        path[0].setPath(left);
        path[1].setPath(right);
    }

    public PathContext(String left, String middle, String right) {
        this();
        files = 3;
        path[0].setPath(left);
        path[1].setPath(middle);
        path[2].setPath(right);
    }

    public PathContext(List<String> paths) {
        this();
        files = clampFileCount(paths.size());
        for (int i = 0; i < files; i++) {
            path[i].setPath(paths.get(i));
        }
    }

    public String getAt(int index) {
        ensureValidIndex(index);
        return path[index].getPath(true);
    }

    public String getElement(int index) {
        ensureValidIndex(index);
        return path[index].getRef();
    }

    public void setAt(int index, String newElement) {
        ensureValidIndex(index);
        path[index].setPath(newElement);
    }

    public String getLeft(boolean normalized) {
        if (files == 0) {
            return "";
        }
        return path[0].getPath(normalized);
    }

    public String getRight(boolean normalized) {
        if (files < 2) {
            return "";
        }
        return path[files - 1].getPath(normalized);
    }

    public String getMiddle(boolean normalized) {
        if (files < 3) {
            return "";
        }
        return path[1].getPath(normalized);
    }

    public String getPath(int index, boolean normalized) {
        ensureArrayIndex(index);
        return path[index].getPath(normalized);
    }

    public void setLeft(String newPath, boolean normalized) {
        if (files == 0) {
            files = 1;
        }
        path[0].setPath(newPath);
        if (normalized) {
            path[0].normalizePath();
        }
    }

    public void setRight(String newPath, boolean normalized) {
        if (files < 2) {
            files = 2;
        }
        path[files - 1].setPath(newPath);
        if (normalized) {
            path[files - 1].normalizePath();
        }
    }

    public void setMiddle(String newPath, boolean normalized) {
        if (files < 3) {
            files = 3;
            path[2].setPath(path[1].getRef());
        }
        path[1].setPath(newPath);
        if (normalized) {
            path[1].normalizePath();
        }
    }

    public void setPath(int index, String newPath, boolean normalized) {
        ensureArrayIndex(index);
        if (index >= files) {
            files = index + 1;
        }
        path[index].setPath(newPath);
        if (normalized) {
            path[index].normalizePath();
        }
    }

    public void setSize(int files) {
        this.files = clampFileCount(files);
    }

    public int getSize() {
        return files;
    }

    public void removeAll() {
        files = 0;
        path[0].setPath("");
        path[1].setPath("");
        path[2].setPath("");
    }

    public void swap(int fromIndex, int toIndex) {
        if (fromIndex >= 0 && fromIndex < files && toIndex >= 0 && toIndex < files) {
            String from = path[fromIndex].getRef();
            path[fromIndex].setPath(path[toIndex].getRef());
            path[toIndex].setPath(from);
        }
    }

    @Override
    public Iterator<String> iterator() {
        List<String> values = new ArrayList<>(files);
        for (int i = 0; i < files; i++) {
            values.add(path[i].getPath(true));
        }
        return values.iterator();
    }

    private void ensureValidIndex(int index) {
        if (index < 0 || index >= files) {
            throw new IllegalArgumentException("index out of range for active path count");
        }
    }

    private void ensureArrayIndex(int index) {
        if (index < 0 || index >= path.length) {
            throw new IllegalArgumentException("index out of range for path array");
        }
    }

    private int clampFileCount(int candidate) {
        return Math.max(0, Math.min(candidate, path.length));
    }
}
